### ISSUE TITLE
Armor stands in ice highway machines act different than in Vanilla

### ISSUE THREAD

**Issue by masterdesky** [PaperMC/Paper#5189]
I was advised on Discord to open a GitHub issue for this since I'm not the first one with this problem, so here we are.

### What behaviour is expected:
On our small 1.16.5 Paper server we've tried to build the following redstone machine, designed by the SciCraft community:
[https://www.youtube.com/watch?v=6Hf62xgulzc](https://www.youtube.com/watch?v=6Hf62xgulzc)
This is a flying machine, which pushes armor stands with frost walker boots on them and simply builds a 4 block wide "highway" with railings on its side as it travels.

The problem arises in the behaviour of the armor stands in the machine. The base concept is, that a slime block is pushed from above into the upper half of each of the armor stands with a piston, and because of this, the flying machine drags them with itself when it moves. It's a somewhat popular design, and it works flawlessly in single player or on Vanilla servers (it was tested multiple times by us too when we kept failing to get it to work on Paper).

### What behaviour is observed:
When the machine is turned on on our Paper server, the armor stands "clip out" from the slime blocks, without moving, completely breaking the machine.

### Steps/models to reproduce:
The build schematic we used can be found in the description of this video:
[https://www.youtube.com/watch?v=5vRmQNTMdco](https://www.youtube.com/watch?v=5vRmQNTMdco)

It is also a block-by-block tutorial, showing the machine in work too.

### Plugin list:
- BetterSleeping 3.0.4
- CoreProtect 19.4
- DynMap 3.1-beta6
- EntityTrackerFixer 1.3.3
- EssentialsX, EssentialsX-Chat, EssentialsX-Spawn 2.18.2.0
- EternalLight 1.7
- LuckPerms 5.2.76
- PlaceholderAPI 2.10.9
- ProtocoLib 4.5.1
- SAML 2.0.0
- ServerListPlus 3.4.8
- SkinsRestorer 13.8.8
- TabList 5.5.0
- Vault 1.7.3
- WorldEdit 7.2.0

### Paper version:
`git-Paper-"1c8b6065e" (MC: 1.16.5)`

### Anything else:
I'm not really sure, but I'm glad to give any more information needed to find any solution for this issue.

**Comment by NNT-Shinso** [comment]
We are having the exact same issue on our server running paper-1.16.4-391. the front of the machine can be "fixed" by adding a slab under the back slime but the railing part of the machine has no push power left over and is unfixable.

as mentioned the armour stands clip out of the slime block they are supposed to be held in.

**Comment by NNT-Shinso** [comment]
No activity on a ticket does not make it solved. The issue remains and still needs to be fixed.

**Comment by electronicboy** [comment]
do you have armor stands left as ticking in paper.yml I think it was? can you reproduce this on spigot?

**Comment by masterdesky** [comment]
I can't speak for others, but in my paper.yml I left `armor-stands-tick` on the default `true` value. As soon as possible I can, I'll test this behaviour on Spigot too. Tbh I didn't test this anywhere else besides Paper since I've opened the issue.

**Comment by NNT-Shinso** [comment]
I have armor-stands-tick on the default true. I have not yet tested on spigot.

**Comment by sausagemountain** [comment]
On spigot armor stands also don't stay in slime blocks

**Comment by petersv5** [comment]
The problem also manifests itself on paper 1.17.1.

**Comment by petersv5** [comment]
Some testing:

All armor stands pulled along with moving (pushed) slime in the same tick always fail or not fail the same. 
If armor stands fail to get pulled along in one tick most of the time subsequent game ticks also have failures. 

If armor stands are forced to move forward by having a solid block behind the slime block they are embedded in they are in fact moved along. However, frosted ice does not form in case this is a "failure tick" as derived from other armor stands failing to get pulled by slime that tick. 

From looking at the symptoms it appears that some ticks have some sort of abbreviated processing where armor stands do not get pulled by the block they are embedded in (but can still be pushed by a solid block behind them) and also frost walker effects are not applied even though an armor stand was pushed.

Let me know if there is any more tests I can perform to help find the cause for this.

**Comment by petersv5** [comment]
The problem can be reproduced with a much simpler setupof:
- in the end dimension or other low light area
-  piston pushing slime-slime-iron
- armor stand whose top half is held inside the middle slime block above and frost walker boots on the armor stand
- 3-wide ice under the armor stand and 3-wide water source blocks in front of the ice.

After a few pushes the armor stand drops back to the slime block closest to the piston and then out alltogether. Sometimes even when the armor stand is pushed frosted ice does not form in from of the armor stand.

**Comment by petersv5** [comment]
After instrumenting the code a bit it seems that in the failing cases there is no call to the handleRelativeFrictionAndCalculateMovement() function in the armor stand after the first tick of extending the piston. 

The moveCollidedEntities() function of the PistonMovingBlockEntity calls the setDeltaMovement() function in the ArmorStand. In the working cases there is a call to handleRelativeFrictionAndCalculateMovement() which I think comes from tick()->aiStep()->travel() in LivingEntity. In the failing case that does not happen, instead I see the next and presumably final tick of the piston extension and no movement is applied.

I am not sure why there is no call to the tick code path yet. Maybe it is an ordering thing? Or can the tick be dropped?

**Comment by petersv5** [comment]
When the armor stand fails to be pulled along the slime block the Armor stand is not ticked (no call to ArmorStand.tick() during the next level tick after the PistonMovingBlockEntity calls setDeltaMovement on the armor stand.

ServerLevel.tickNonPassenger() _is_ called on the armor stand entity at the correct time, but that does not lead to a call to ArmorStand.tick() during the failures.

**Comment by petersv5** [comment]
The root problem seems to be the throwing away of 1/4 of all ticks for entities in org.spigotmc.ActivationRange(), near the end. This will impact all tight timing based designs, and especially piston based movements of entities. 

By disabling this feature the ice maker works, as do all smaller scale tests I have run. This solves both the armor stands not being pulled and the lack of frosted ice making.

I suppose the immunity system could be modified to apply to all entities that are involved in a piston movement. I am not sure if it should be triggered by the entity being inside a block or but a push happening or sometihng else.

**Comment by petersv5** [comment]
I have two patches, one the adds a very targeted piston induced immunity by adding a variable in Entity that remmebers that it was affected by a PistonMovingEntity, and which is cleared by the next Entity.tick(). The logic about how many ticks need immunity from the decimation is handled inside ActivationRange. 

The second instead adds a memory of the amount of ticks the entity needs to be immune from tick decimation, decremented by Entity.tick(). ActvationRange thus does not have to know about all cases where an Entity may have to have all ticks, it just asks the Entity. 

I am not sure which of these should be submitted or if there is some third way that is preferred.

**Comment by qixils** [comment]
thank you for your work on this peters :) you could submit both and let maintainers accept which they think is best?

**Comment by petersv5** [comment]
I have created pull request 6301 with one of the two fixes. The suggestion on discord paper-dev was to just submit one of them and mention the other alternative.

I ran the Ilmango contraption for 2000 blocks. There is an issue with endermen. It seems that if they are able to spawn on the machine (i.e. no player is close enough to prevent spawning on the blocks) and they then fall into the water it does not form frosted ice. I think this is normal vanilla behviour, but otherwise I will open a new ticket. It is anyway no showstopper, you just have to kill off all endermen before starting the machine.

**Comment by NNT-Shinso** [comment]
> 
> 
> I have created pull request 6301 with one of the two fixes. The suggestion on discord paper-dev was to just submit one of them and mention the other alternative.
> 
> I ran the Ilmango contraption for 2000 blocks. There is an issue with endermen. It seems that if they are able to spawn on the machine (i.e. no player is close enough to prevent spawning on the blocks) and they then fall into the water it does not form frosted ice. I think this is normal vanilla behviour, but otherwise I will open a new ticket. It is anyway no showstopper, you just have to kill off all endermen before starting the machine.

Thanks a lot for checking this. And the endermen is just as you say, kill them before starting and ride the machine so they don't spawn nearby.

**Comment by qixils** [comment]
yes, it is intended vanilla behavior that frosted ice cannot form if a mob is inside the water

**Comment by petersv5** [comment]
I managed to close the PR accidentally. I'll create a new, fixed one. Sorry for the inconvenience.

**Comment by petersv5** [comment]
New PR #6308 created, now including the missing file.

**Comment by masterdesky** [comment]
Thank you @petersv5 for your throughout testing and for the fix!

**Comment by petersv5** [comment]
The Discord was helpful in pointing out areas that needed checking and verifying.
Pity I did not verify that the first patch I submitted included all changed files. :-(
Fixed not though.
